content/200-concepts/100-components/02-prisma-client/260-client-extensions/120-query.mdx

```diff
@@ -1,16 +1,186 @@
+
+In the above example, a call to `xprisma.user.findMany` triggers `query.user.findMany`. Each callback receives a type-safe `{ model, operation, args, data }` object that describes the query. This object has the following properties:
+
+- `model`: the name of the containing model for the query that we want to extend.
+
+  In the above example, the `model` is a string of type `"User"`.
+
+- `operation`: the name of the operation being extended and executed.
+
+  In the above example, the `operation` is a string of type `"findMany"`.
+
+  ** TBC: should this be `operation` or `action`? **
+
+- `args`: the specific query input information to be extended.
+
+  This is a type-safe object that you can mutate before the query happens. You can mutate any of the properties in `args`. Exception: you cannot mutate `include` or `select` because that would change the expected output type and break type safety.
+
+- `data`: a promise for the result of the query.
+
+  - You can use `await` and then mutate this promise, because its value is type-safe. TypeScript catches any unsafe mutations on the object.
+  - `await data` is equivalent to calling `next(...)` in a middleware.
+
+### Add a custom query to a specific operation in all models of your schema
+
+To extend all models in your schema, use `$allmodels` instead of a specific model name. For example:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    user: {
+      async findMany({ model, action, args, data }) {
+        // `where` can be `undefined`, so `args.where.age` might
+        // fail. To work around this, we assign an object and
+        // backfill the rest of the `where` clause:
+        args.where = { age: { gt: 18 }, ...args.where }
+        const users = await data
+        return data
+      },
+    },
+  },
+})
+```
+
+Note: recursive wildcards just modify the query itself. They cannot return values.
+
+** TBD - can we have an example of this note? **
+
+### Add a custom query to all operations in a specific model
+
+Use `$allOperations` to apply your custome query to all operations, instead of to a specific operation name.
+
+For example, the following code applies a custom query to all operations on the `user` model:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    user: {
+      $allOperations({ model, operation, args, data }) {
+        // handle all operations
+        return data
+      },
+    },
+  },
+})
+```
+
+** TBD: are the parameters model, operation, args, data or model, action, args, data? **
+
+### Add a custom query to all operations in all models of your schema
+
+Use `$allOperations` to apply your custome query to all operations, instead of to a specific operation name.
+
+To apply a custom query to all operations on all models of your schema:
+
+```ts
+const xprisma = prisma.$extends({
+  query: {
+    $allModels: {
+      $allOperations({ model, operation, args, data }) {
```

Comment:
Are the parameters here:
- `model`, `operation`, `args`, `data` or 
- `model`, `action`, `args`, `data`?

Reply:
Operation